// @sumup/foundry/dist/types/shared.js
"use strict";
/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Framework = exports.Environment = exports.Language = exports.Prompt = exports.Tool = exports.Preset = void 0;
var Preset;
(function (Preset) {
    Preset["LINT"] = "lint";
})(Preset = exports.Preset || (exports.Preset = {}));
var Tool;
(function (Tool) {
    Tool["ESLINT"] = "eslint";
    Tool["STYLELINT"] = "stylelint";
    Tool["PRETTIER"] = "prettier";
    Tool["HUSKY"] = "husky";
    Tool["LINT_STAGED"] = "lint-staged";
})(Tool = exports.Tool || (exports.Tool = {}));
var Prompt;
(function (Prompt) {
    Prompt["OPEN_SOURCE"] = "open-source";
})(Prompt = exports.Prompt || (exports.Prompt = {}));
var Language;
(function (Language) {
    Language["TYPESCRIPT"] = "TypeScript";
    Language["JAVASCRIPT"] = "JavaScript";
})(Language = exports.Language || (exports.Language = {}));
var Environment;
(function (Environment) {
    Environment["NODE"] = "Node";
    Environment["BROWSER"] = "Browser";
})(Environment = exports.Environment || (exports.Environment = {}));
var Framework;
(function (Framework) {
    Framework["REACT"] = "React";
    Framework["NEXT_JS"] = "Next.js";
    Framework["EMOTION"] = "Emotion";
    Framework["JEST"] = "Jest";
    Framework["TESTING_LIBRARY"] = "Testing Library";
    Framework["CYPRESS"] = "Cypress";
    Framework["PLAYWRIGHT"] = "Playwright";
})(Framework = exports.Framework || (exports.Framework = {}));
