// audit-ci/dist/common.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deduplicate = exports.gitHubAdvisoryIdToUrl = exports.gitHubAdvisoryUrlToAdvisoryId = exports.isGitHubAdvisoryId = exports.matchString = exports.runProgram = exports.reportAudit = exports.partition = void 0;
const cross_spawn_1 = require("cross-spawn");
const escape_string_regexp_1 = __importDefault(require("escape-string-regexp"));
const eventStream = __importStar(require("event-stream"));
const JSONStream = __importStar(require("JSONStream"));
const readline_transform_1 = __importDefault(require("readline-transform"));
const colors_1 = require("./colors");
function partition(a, fun) {
    const returnValue = { truthy: [], falsy: [] };
    for (const item of a) {
        if (fun(item)) {
            returnValue.truthy.push(item);
        }
        else {
            returnValue.falsy.push(item);
        }
    }
    return returnValue;
}
exports.partition = partition;
function reportAudit(summary, config) {
    const { allowlist, "show-not-found": showNotFound, "show-found": showFound, "output-format": outputFormat, } = config;
    const { allowlistedModulesFound, allowlistedAdvisoriesFound, allowlistedModulesNotFound, allowlistedAdvisoriesNotFound, allowlistedPathsNotFound, failedLevelsFound, advisoriesFound, advisoryPathsFound, } = summary;
    if (outputFormat === "text") {
        if (allowlist.modules.length > 0) {
            console.log(colors_1.blue, `Modules to allowlist: ${allowlist.modules.join(", ")}.`);
        }
        if (showFound) {
            if (allowlistedModulesFound.length > 0) {
                const found = allowlistedModulesFound.join(", ");
                console.warn(colors_1.yellow, `Found vulnerable allowlisted modules: ${found}.`);
            }
            if (allowlistedAdvisoriesFound.length > 0) {
                const found = allowlistedAdvisoriesFound.join(", ");
                console.warn(colors_1.yellow, `Found vulnerable allowlisted advisories: ${found}.`);
            }
        }
        if (showNotFound) {
            if (allowlistedModulesNotFound.length > 0) {
                const found = allowlistedModulesNotFound
                    .sort((a, b) => a.localeCompare(b))
                    .join(", ");
                const allowlistMessage = allowlistedModulesNotFound.length === 1
                    ? `Consider not allowlisting module: ${found}.`
                    : `Consider not allowlisting modules: ${found}.`;
                console.warn(colors_1.yellow, allowlistMessage);
            }
            if (allowlistedAdvisoriesNotFound.length > 0) {
                const found = allowlistedAdvisoriesNotFound
                    .sort((a, b) => a.localeCompare(b))
                    .join(", ");
                const allowlistMessage = allowlistedAdvisoriesNotFound.length === 1
                    ? `Consider not allowlisting advisory: ${found}.`
                    : `Consider not allowlisting advisories: ${found}.`;
                console.warn(colors_1.yellow, allowlistMessage);
            }
            if (allowlistedPathsNotFound.length > 0) {
                const found = allowlistedPathsNotFound
                    .sort((a, b) => a.localeCompare(b))
                    .join(", ");
                const allowlistMessage = allowlistedPathsNotFound.length === 1
                    ? `Consider not allowlisting path: ${found}.`
                    : `Consider not allowlisting paths: ${found}.`;
                console.warn(colors_1.yellow, allowlistMessage);
            }
        }
        if (advisoryPathsFound.length > 0) {
            const found = advisoryPathsFound.join("\n");
            console.warn(colors_1.yellow, `Found vulnerable advisory paths:`);
            console.log(found);
        }
    }
    if (failedLevelsFound.length > 0) {
        // Get the levels that have failed by filtering the keys with true values
        throw new Error(`Failed security audit due to ${failedLevelsFound.join(", ")} vulnerabilities.\nVulnerable advisories are:\n${advisoriesFound
            .map((element) => gitHubAdvisoryIdToUrl(element))
            .join("\n")}`);
    }
    return summary;
}
exports.reportAudit = reportAudit;
function hasMessage(value) {
    return typeof value === "object" && value != undefined && "message" in value;
}
function hasStatusCode(value) {
    return (typeof value === "object" && value != undefined && "statusCode" in value);
}
function runProgram(command, arguments_, options, stdoutListener, stderrListener) {
    const transform = new readline_transform_1.default({ skipEmpty: true });
    const proc = (0, cross_spawn_1.spawn)(command, arguments_, options);
    let recentMessage;
    let errorMessage;
    proc.stdout.setEncoding("utf8");
    proc.stdout
        .pipe(transform.on("error", (error) => {
        throw error;
    }))
        .pipe(eventStream.mapSync((data) => {
        recentMessage = data;
        return data;
    }))
        .pipe(
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore -- JSONStream.parse() accepts (pattern: any) when it should accept (pattern?: any)
    JSONStream.parse().on("error", () => {
        errorMessage = recentMessage;
        throw new Error(errorMessage);
    }))
        .pipe(eventStream.mapSync((data) => {
        if (!data)
            return;
        try {
            // due to response without error
            if (hasMessage(data) &&
                typeof data.message === "string" &&
                data.message.includes("ENOTFOUND")) {
                stderrListener(data.message);
                return;
            }
            // TODO: There are no tests that cover this case, not sure when this happens.
            if (hasStatusCode(data) && data.statusCode === 404) {
                stderrListener(data.message);
                return;
            }
            stdoutListener(data);
        }
        catch (error) {
            stderrListener(error);
        }
    }));
    return new Promise((resolve, reject) => {
        proc.on("close", () => {
            if (errorMessage) {
                return reject(new Error(errorMessage));
            }
            return resolve();
        });
        proc.on("error", (error) => reject(errorMessage ? new Error(errorMessage) : error));
    });
}
exports.runProgram = runProgram;
function wildcardToRegex(stringWithWildcard) {
    const regexString = stringWithWildcard
        .split(/\*+/) // split at every wildcard (*) character
        .map((s) => (0, escape_string_regexp_1.default)(s)) // escape the substrings to make sure that they aren't evaluated
        .join(".*"); // construct a regex matching anything at each wildcard location
    return new RegExp(`^${regexString}$`);
}
function matchString(template, string_) {
    return template.includes("*")
        ? wildcardToRegex(template).test(string_)
        : template === string_;
}
exports.matchString = matchString;
function isGitHubAdvisoryId(id) {
    return typeof id === "string" && id.startsWith("GHSA");
}
exports.isGitHubAdvisoryId = isGitHubAdvisoryId;
function gitHubAdvisoryUrlToAdvisoryId(url) {
    return url.split("/")[4];
}
exports.gitHubAdvisoryUrlToAdvisoryId = gitHubAdvisoryUrlToAdvisoryId;
function gitHubAdvisoryIdToUrl(id) {
    return `https://github.com/advisories/${id}`;
}
exports.gitHubAdvisoryIdToUrl = gitHubAdvisoryIdToUrl;
function deduplicate(array) {
    return [...new Set(array)];
}
exports.deduplicate = deduplicate;
