// audit-ci/dist/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runYargs = void 0;
const fs_1 = require("fs");
const jju_1 = require("jju");
// eslint-disable-next-line unicorn/import-style
const path = __importStar(require("path"));
const yargs_1 = require("yargs");
const allowlist_1 = __importDefault(require("./allowlist"));
const map_vulnerability_1 = require("./map-vulnerability");
function mapReportTypeInput(config) {
    const { "report-type": reportType } = config;
    switch (reportType) {
        case "full":
        case "important":
        case "summary":
            return reportType;
        default:
            throw new Error(`Invalid report type: ${reportType}. Should be \`['important', 'full', 'summary']\`.`);
    }
}
function mapExtraArgumentsInput(config) {
    // These args will often be flags for another command, so we
    // want to have some way of escaping args that start with a -.
    // We'll look for and remove a single backslash at the start, if present.
    return config["extra-args"].map((a) => a.replace(/^\\/, ""));
}
/**
 * @param pmArgument the package manager (including the `auto` option)
 * @param directory the directory where the package manager files exist
 * @returns the non-`auto` package manager
 */
function resolvePackageManagerType(pmArgument, directory) {
    switch (pmArgument) {
        case "npm":
        case "pnpm":
        case "yarn":
            return pmArgument;
        case "auto": {
            const getPath = (file) => path.resolve(directory, file);
            const packageLockExists = (0, fs_1.existsSync)(getPath("package-lock.json"));
            if (packageLockExists)
                return "npm";
            const shrinkwrapExists = (0, fs_1.existsSync)(getPath("npm-shrinkwrap.json"));
            if (shrinkwrapExists)
                return "npm";
            const yarnLockExists = (0, fs_1.existsSync)(getPath("yarn.lock"));
            if (yarnLockExists)
                return "yarn";
            const pnpmLockExists = (0, fs_1.existsSync)(getPath("pnpm-lock.yaml"));
            if (pnpmLockExists)
                return "pnpm";
            throw new Error("Cannot establish package-manager type, missing package-lock.json, yarn.lock, and pnpm-lock.yaml.");
        }
        default:
            throw new Error(`Unexpected package manager argument: ${pmArgument}`);
    }
}
function mapArgvToAuditCiConfig(argv) {
    const allowlist = allowlist_1.default.mapConfigToAllowlist(argv);
    const { low, moderate, high, critical, "package-manager": packageManager, directory, } = argv;
    const resolvedPackageManager = resolvePackageManagerType(packageManager, directory);
    const result = {
        ...argv,
        "package-manager": resolvedPackageManager,
        levels: (0, map_vulnerability_1.mapVulnerabilityLevelInput)({
            low,
            moderate,
            high,
            critical,
        }),
        "report-type": mapReportTypeInput(argv),
        allowlist: allowlist,
        "extra-args": mapExtraArgumentsInput(argv),
    };
    return result;
}
async function runYargs() {
    const { argv } = (0, yargs_1.config)("config", (configPath) => 
    // Supports JSON, JSONC, & JSON5
    (0, jju_1.parse)((0, fs_1.readFileSync)(configPath, "utf8"), {
        // When passing an allowlist using NSRecord syntax, yargs will throw an error
        // "Invalid JSON config file". We need to add this flag to prevent that.
        null_prototype: false,
    }))
        .options({
        l: {
            alias: "low",
            default: false,
            describe: "Exit for low vulnerabilities or higher",
            type: "boolean",
        },
        m: {
            alias: "moderate",
            default: false,
            describe: "Exit for moderate vulnerabilities or higher",
            type: "boolean",
        },
        h: {
            alias: "high",
            default: false,
            describe: "Exit for high vulnerabilities or higher",
            type: "boolean",
        },
        c: {
            alias: "critical",
            default: false,
            describe: "Exit for critical vulnerabilities",
            type: "boolean",
        },
        p: {
            alias: "package-manager",
            default: "auto",
            describe: "Choose a package manager",
            choices: ["auto", "npm", "yarn", "pnpm"],
        },
        r: {
            alias: "report",
            default: false,
            describe: "Show a full audit report",
            type: "boolean",
        },
        s: {
            alias: "summary",
            default: false,
            describe: "Show a summary audit report",
            type: "boolean",
        },
        a: {
            alias: "allowlist",
            default: [],
            describe: "Allowlist module names (example), advisories (123), and module paths (123|example1>example2)",
            type: "array",
        },
        d: {
            alias: "directory",
            default: "./",
            describe: "The directory containing the package.json to audit",
            type: "string",
        },
        o: {
            alias: "output-format",
            default: "text",
            describe: "The format of the output of audit-ci",
            choices: ["text", "json"],
        },
        "show-found": {
            default: true,
            describe: "Show allowlisted advisories that are found",
            type: "boolean",
        },
        "show-not-found": {
            default: true,
            describe: "Show allowlisted advisories that are not found",
            type: "boolean",
        },
        registry: {
            default: undefined,
            describe: "The registry to resolve packages by name and version",
            type: "string",
        },
        "report-type": {
            default: "important",
            describe: "Format for the audit report results",
            type: "string",
            choices: ["important", "summary", "full"],
        },
        "retry-count": {
            default: 5,
            describe: "The number of attempts audit-ci calls an unavailable registry before failing",
            type: "number",
        },
        "pass-enoaudit": {
            default: false,
            describe: "Pass if no audit is performed due to the registry returning ENOAUDIT",
            type: "boolean",
        },
        "skip-dev": {
            default: false,
            describe: "Skip devDependencies",
            type: "boolean",
        },
        "extra-args": {
            default: [],
            describe: "Pass additional arguments to the underlying audit command",
            type: "array",
        },
    })
        .help("help");
    // yargs doesn't support aliases + TypeScript
    const awaitedArgv = (await argv);
    const auditCiConfig = mapArgvToAuditCiConfig(awaitedArgv);
    return auditCiConfig;
}
exports.runYargs = runYargs;
