// audit-ci/dist/yarn-auditer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.audit = void 0;
const child_process_1 = require("child_process");
const semver = __importStar(require("semver"));
const colors_1 = require("./colors");
const common_1 = require("./common");
const model_1 = __importDefault(require("./model"));
const MINIMUM_YARN_CLASSIC_VERSION = "1.12.3";
const MINIMUM_YARN_BERRY_VERSION = "2.4.0";
/**
 * Change this to the appropriate version when
 * yarn audit --registry is supported:
 * @see https://github.com/yarnpkg/yarn/issues/7012
 */
const MINIMUM_YARN_AUDIT_REGISTRY_VERSION = "99.99.99";
function getYarnVersion(cwd) {
    const version = (0, child_process_1.execSync)("yarn -v", { cwd }).toString().replace("\n", "");
    return version;
}
function yarnSupportsClassicAudit(yarnVersion) {
    return semver.satisfies(yarnVersion, `^${MINIMUM_YARN_CLASSIC_VERSION}`);
}
function yarnSupportsBerryAudit(yarnVersion) {
    return semver.gte(yarnVersion, MINIMUM_YARN_BERRY_VERSION);
}
function yarnSupportsAudit(yarnVersion) {
    return (yarnSupportsClassicAudit(yarnVersion) || yarnSupportsBerryAudit(yarnVersion));
}
function yarnAuditSupportsRegistry(yarnVersion) {
    return semver.gte(yarnVersion, MINIMUM_YARN_AUDIT_REGISTRY_VERSION);
}
const printJson = (data) => {
    console.log(JSON.stringify(data, undefined, 2));
};
/**
 * Audit your Yarn project!
 *
 * @returns Returns the audit report summary on resolve, `Error` on rejection.
 */
async function audit(config, reporter = common_1.reportAudit) {
    const { levels, registry, "report-type": reportType, "skip-dev": skipDevelopmentDependencies, "output-format": outputFormat, _yarn, directory, "extra-args": extraArguments, } = config;
    const yarnExec = _yarn || "yarn";
    let missingLockFile = false;
    const model = new model_1.default(config);
    const yarnVersion = getYarnVersion(directory);
    const isYarnVersionSupported = yarnSupportsAudit(yarnVersion);
    if (!isYarnVersionSupported) {
        throw new Error(`Yarn ${yarnVersion} not supported, must be ^${MINIMUM_YARN_CLASSIC_VERSION} or >=${MINIMUM_YARN_BERRY_VERSION}`);
    }
    const isYarnClassic = yarnSupportsClassicAudit(yarnVersion);
    const yarnName = isYarnClassic ? `Yarn` : `Yarn Berry`;
    function isClassicGuard(response) {
        return isYarnClassic;
    }
    const printHeader = (text) => {
        if (outputFormat === "text") {
            console.log(colors_1.blue, text);
        }
    };
    switch (reportType) {
        case "full":
            printHeader(`${yarnName} audit report JSON:`);
            break;
        case "important":
            printHeader(`${yarnName} audit report results:`);
            break;
        case "summary":
            printHeader(`${yarnName} audit report summary:`);
            break;
        default:
            throw new Error(`Invalid report type: ${reportType}. Should be \`['important', 'full', 'summary']\`.`);
    }
    // Define a function to print based on the report type.
    let printAuditData;
    switch (reportType) {
        case "full":
            printAuditData = (line) => {
                printJson(line);
            };
            break;
        case "important":
            printAuditData = isYarnClassic
                ? ({ type, data }) => {
                    if ((type === "auditAdvisory" && levels[data.advisory.severity]) ||
                        type === "auditSummary") {
                        printJson(data);
                    }
                }
                : ({ metadata }) => {
                    printJson(metadata);
                };
            break;
        case "summary":
            printAuditData = isYarnClassic
                ? ({ type, data }) => {
                    if (type === "auditSummary") {
                        printJson(data);
                    }
                }
                : ({ metadata }) => {
                    printJson(metadata);
                };
            break;
        default:
            throw new Error(`Invalid report type: ${reportType}. Should be \`['important', 'full', 'summary']\`.`);
    }
    function outListener(line) {
        try {
            if (isClassicGuard(line)) {
                const { type, data } = line;
                printAuditData(line);
                if (type === "info" && data === "No lockfile found.") {
                    missingLockFile = true;
                    return;
                }
                if (type !== "auditAdvisory") {
                    return;
                }
                model.process(data.advisory);
            }
            else {
                printAuditData(line);
                if ("advisories" in line) {
                    for (const advisory of Object.values(line.advisories)) {
                        model.process(advisory);
                    }
                }
            }
        }
        catch (error) {
            console.error(colors_1.red, `ERROR: Cannot JSONStream.parse response:`);
            console.error(line);
            throw error;
        }
    }
    const stderrBuffer = [];
    function errorListener(line) {
        stderrBuffer.push(line);
        if (line.type === "error") {
            throw new Error(line.data);
        }
    }
    const options = { cwd: directory };
    const arguments_ = isYarnClassic
        ? [
            "audit",
            "--json",
            ...(skipDevelopmentDependencies ? ["--groups", "dependencies"] : []),
        ]
        : [
            "npm",
            "audit",
            "--recursive",
            "--json",
            "--all",
            ...(skipDevelopmentDependencies ? ["--environment", "production"] : []),
        ];
    if (registry) {
        const auditRegistrySupported = yarnAuditSupportsRegistry(yarnVersion);
        if (auditRegistrySupported) {
            arguments_.push("--registry", registry);
        }
        else {
            console.warn(colors_1.yellow, "Yarn audit does not support the registry flag yet.");
        }
    }
    if (extraArguments) {
        arguments_.push(...extraArguments);
    }
    await (0, common_1.runProgram)(yarnExec, arguments_, options, outListener, errorListener);
    if (missingLockFile) {
        console.warn(colors_1.yellow, "No yarn.lock file. This does not affect auditing, but it may be a mistake.");
    }
    const summary = model.getSummary((a) => a.github_advisory_id);
    return reporter(summary, config);
}
exports.audit = audit;
